File: Graph/FloodFill/fill.cpp
#include <vector>
#include <queue>
using namespace std;

class Solution {
  public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc,
                                  int newColor) {
        // code here
        int m = image.size();
        int n = image[0].size();
        queue<vector<int>>qu;
        vector<vector<bool>>visited(m,vector<bool>(n,false));
        qu.push({sr,sc});
        int origColor = image[sr][sc];
        image[sr][sc] = newColor;
        visited[sr][sc] = true;
        int dx[4] = {-1,0,1,0};
        int dy[4] = {0,-1,0,1};
        while(!qu.empty()){
            auto curr = qu.front();
            qu.pop();
            int r = curr[0];
            int c = curr[1];
            for(int i=0;i<4;i++){
                int nr = r+dx[i];
                int nc = c+dy[i];
                if(nr>=0 && nr<m && nc>=0 && nc<n){
                    if(!visited[nr][nc] && image[nr][nc]==origColor){
                        qu.push({nr,nc});
                        image[nr][nc] = newColor;
                        visited[nr][nc] = true;
                    }
                }
            }
        }
        return image;
    }
};

File: Graph/FloodFill/fill.py
from collections import deque

class Solution:
    def floodFill(self, image, sr, sc, newColor):
        m = len(image)
        n = len(image[0])

        qu = deque()
        visited = [[False] * n for _ in range(m)]

        qu.append((sr, sc))
        origColor = image[sr][sc]
        image[sr][sc] = newColor
        visited[sr][sc] = True

        dx = [-1, 0, 1, 0]
        dy = [0, -1, 0, 1]

        while qu:
            r, c = qu.popleft()

            for i in range(4):
                nr = r + dx[i]
                nc = c + dy[i]

                if 0 <= nr < m and 0 <= nc < n:
                    if (not visited[nr][nc] and
                        image[nr][nc] == origColor):

                        qu.append((nr, nc))
                        image[nr][nc] = newColor
                        visited[nr][nc] = True

        return image